spectec/src/middlend/uncaseremoval.ml

```diff
@@ -0,0 +1,383 @@
+  def $proj_foo_0{var : foo}(var) = ?()
+
+with uncase expressions being transformed into:
+(the($proj_foo_0(x))).0 and (the($proj_foo_1(x))).1
+
+Names were specifically chosen here for simplicity.
+*)
+
+open Il.Ast
+open Il
+open Util.Source
+
+module StringMap = Map.Make(String)
+type uncase_map = (mixop list) StringMap.t
+
+type env = {
+  mutable uncase_map : uncase_map;
+  mutable env : Il.Env.t;
+}
+
+let empty_env = {
+  uncase_map = StringMap.empty;
+  env = Il.Env.empty
+}
+
+let error at msg = Util.Error.error at "uncase-removal" msg
+
+let var_prefix = "v_"
+let proj_prefix = "proj_"
+
+let get_case_typs t = 
+  match t.it with
+  | TupT typs -> typs
+  | _ -> [(VarE ("_" $ t.at) $$ t.at % t, t)]
+
+let iter_name i = 
+  match i with
+  | Opt -> "opt"
+  | List | List1 | ListN _ -> "list"
+
+let rec typ_name t =
+  match t.it with
+  | VarT (id, _) -> id.it
+  | BoolT -> "bool"
+  | NumT _ -> "num"
+  | TextT -> "str"
+  | TupT pairs -> String.concat "__" (List.map (fun (_, t) -> typ_name t) pairs) ^ "_pair"
+  | IterT (typ, iter) -> typ_name typ ^ "_" ^ iter_name iter 
+
+let make_arg p = 
+  (match p.it with
+  | ExpP (id, typ) -> ExpA (VarE id $$ id.at % typ) 
+  | TypP id -> TypA (VarT (id, []) $ id.at) (* TODO unsure this makes sense*)
+  | DefP (id, _, _) -> DefA id 
+  | GramP (_, _) -> assert false (* Avoid this *)
+  ) $ p.at
+
+let make_bind p = 
+  (match p.it with 
+  | ExpP (id, typ) -> ExpB (id, typ)
+  | TypP id -> TypB id
+  | DefP (id, params, typ) -> DefB (id, params, typ)
+  | GramP _ -> assert false (* Avoid this *)
+  ) $ p.at
+
+let create_projection_functions id params mixops inst =
+  let get_deftyp inst' = (match inst'.it with
+    | InstD (_binds, _args, deftyp) -> deftyp.it
+  ) in 
+  let at = inst.at in 
+  let user_typ = VarT (id, List.map make_arg params) $ at in 
+  let param_ids = List.map (fun p -> (Utils.get_param_id p).it) params in 
+  let fresh_name = Utils.generate_var param_ids "x" in
+  let new_param = ExpP (fresh_name $ at, user_typ) $ at in
+  let make_func m case_typs has_one_case idx = 
+    let new_params = params @ [new_param] in 
+    let new_var_exps = List.mapi (fun idx (_, t) -> VarE (var_prefix ^ typ_name t ^ "_" ^ Int.to_string idx $ at) $$ at % t) case_typs in 
+    let tupt = TupT case_typs $ at in
+    let no_name_tupt = TupT (List.map (fun (e, t) -> ({e with it = VarE ("_" $ e.at)}, t)) case_typs) $ at in
+    let new_tup = TupE (new_var_exps) $$ at % tupt in
+    let new_case_exp = CaseE(m, new_tup) $$ at % user_typ in
+    let new_arg = ExpA new_case_exp $ at in 
+    let new_binds = List.mapi (fun idx (_, t) -> ExpB (var_prefix ^ typ_name t ^ "_" ^ Int.to_string idx $ at, t) $ at) case_typs in
+    let clause = DefD (List.map make_bind params @ new_binds, List.map make_arg params @ [new_arg], new_tup, []) $ at in 
+    let normal = DecD ((proj_prefix ^ id.it ^ "_" ^ Int.to_string idx) $ id.at, new_params, no_name_tupt, [clause]) in
+
+    if has_one_case then normal else
```

Comment:
Just a nit, but (unless you go via the partial hint anyways), can you `if has_one_case` earlier, and indent, so that those bindings that are only relevant in the then-case are really only in the the then case?

```
let make_func …
  let …
  if has_one_case then
    let …
    DecD …
  else
    let …
    DecD …
```